_stbt/imgproc_cache.py

```diff
@@ -0,0 +1,211 @@
+from _stbt.logging import ImageLogger
+from _stbt.utils import mkdir_p
+
+_cache = None
+
+
+@contextmanager
+def cache(filename=None):
+    if os.environ.get('STBT_DISABLE_CACHING'):
+        yield
+        return
+
+    global _cache
+
+    sys.path.append(os.path.dirname(__file__))
+    import _stbt.lmdb as lmdb
+    del sys.path[-1]
+
+    if filename is None:
+        cache_home = os.environ.get('XDG_CACHE_HOME') \
+            or '%s/.cache' % os.environ['HOME']
+        mkdir_p(cache_home + "/stbt")
+        filename = cache_home + "/stbt/cache.lmdb"
+    with lmdb.open(filename) as db:  # pylint: disable=no-member
+        assert _cache is None
+        try:
+            _cache = db
+            yield
+        finally:
+            _cache = None
+
+
+def memoize(function):
+    funcname = function.__name__
+
+    @functools.wraps(function)
+    def inner(*args, **kwargs):
+        try:
+            if _cache is None:
+                raise NotCachable()
+            full_kwargs = inspect.getcallargs(function, *args, **kwargs)
+            key = _cache_hash((funcname, full_kwargs))
+            with _cache.begin() as txn:
+                out = txn.get(key)
+            if out is not None:
+                return json.loads(out)
+            output = function(**full_kwargs)
+            with _cache.begin(write=True) as txn:
+                out = txn.put(key, json.dumps(output).encode("utf-8"))
+            return output
+        except NotCachable:
+            return function(*args, **kwargs)
+
+    return inner
+
+
+class NotCachable(Exception):
+    pass
+
+
+class _ArgsEncoder(json.JSONEncoder):
+    def default(self, value):  # pylint: disable=method-hidden
+        import _stbt.core as core
+        if isinstance(value, ImageLogger):
+            if value.enabled:
+                raise NotCachable()
+            return None
+        elif isinstance(value, set):
+            return sorted(value)
+        elif isinstance(value, core.MatchParameters):
+            return {
+                "match_method": value.match_method,
+                "match_threshold": value.match_threshold,
+                "confirm_method": value.confirm_method,
+                "confirm_threshold": value.confirm_threshold,
+                "erode_passes": value.erode_passes}
+        elif isinstance(value, (Gst.Sample, numpy.ndarray)):
+            from _stbt.xxhash import Xxhash64
+            with numpy_from_sample(value, readonly=True) as s:
+                h = Xxhash64()
+                h.update(numpy.ascontiguousarray(s).data)
+                return (s.shape, h.hexdigest())
+        else:
+            json.JSONEncoder.default(self, value)
+
+
+def _cache_hash(value):
+    from _stbt.xxhash import Xxhash64
+    h = Xxhash64()
+
+    class HashWriter(object):
+        def write(self, data):
+            h.update(data)
+            return len(data)
+
+    json.dump(value, HashWriter(), cls=_ArgsEncoder)
```

Comment:
And another issue: we need to have `sort_keys=True` here to make it deterministic.